pkg/ccl/streamingccl/logical/lww_row_processor.go

```diff
@@ -156,12 +185,32 @@ func (lww *sqlLastWriteWinsRowProcessor) insertRow(
 	if !ok {
 		return errors.Errorf("no pre-generated insert query for table %d", row.TableID)
 	}
-	insertQuery, ok := insertQueriesForTable[row.FamilyID]
+	insertQueries, ok := insertQueriesForTable[row.FamilyID]
 	if !ok {
 		return errors.Errorf("no pre-generated insert query for table %d column family %d", row.TableID, row.FamilyID)
 	}
-	if _, err := txn.ExecParsed(ctx, "replicated-insert", txn.KV(), attributeToUser, insertQuery, datums...); err != nil {
-		log.Warningf(ctx, "replicated insert failed (query: %s): %s", insertQuery.SQL, err.Error())
+	if prevValue.RawBytes == nil && tryOptimisticInsertEnabled.Get(&lww.settings.SV) {
+		// If there was no value before this change on the source, it's likely
+		// that the original stmt was an insert that didn't hit a conflict, so
+		// we'll try the optimistic insert too.
+		query := insertQueries[insertQueriesOptimisticIndex]
+		if _, err = txn.ExecParsed(ctx, "replicated-optimistic-insert", txn.KV(), attributeToUser, query, datums...); err != nil {
+			// If the optimistic insert failed with unique violation, we have to
+			// fall back to the pessimistic path. If we got a different error,
+			// then we bail completely.
+			if pgerror.GetPGCode(err) != pgcode.UniqueViolation {
+				log.Warningf(ctx, "replicated optimistic insert failed (query: %s): %s", query.SQL, err.Error())
+				return err
+			}
```

Comment:
Might not be worth it, but if we wanted we could perhaps use pgerror.GetConstraintName() and check if the conflict is on the primary index as well.